*: Upstream parity-multiaddr changes

mxinden opened the pull request:
[`parity-multiaddr`](https://github.com/libp2p/rust-libp2p/tree/master/misc/multiaddr) is a fork of this repository with various changes and additions. As part of https://github.com/libp2p/rust-libp2p/issues/758 this pull request does the following in order to merge the two while preserving the git history of both repositories:

- Use `git subtree split` to extract https://github.com/libp2p/rust-libp2p/tree/master/misc/multiaddr to a separate git repository.
- Rebase extracted commits onto f34de5ce4cb4a25296dc4affa30b1d528a96a1a5 in order for them to share a common commit.


Main merge happens via https://github.com/multiformats/rust-multiaddr/pull/40/commits/935a9664b4b12048fc0babbe734e69923b489a28. As far as I can tell all features of `multiaddr` and `parity-multiaddr` are preserved. Naming closely follows `parity-multiaddr`, thus crates depending on `multiaddr` (as far as I can tell [none](https://crates.io/crates/multiaddr/reverse_dependencies)) will face breaking changes. In case of conflicts `parity-multiaddr` is favored over `multiaddr` given that the former is more up-to-date and has a larger user base. Version is fast-forwarded to `v0.12.0` to not conflict with existing `parity-multiaddr` versions.

mxinden commented:
@vmx @Stebalien does one of you have some time to give this pull request a review? I am sorry for the rather large change-set. Let me know in case you see a way to simplify the review work.

vmx reviewed (commented):
As the parity-multiaddr changes were already reviewed, I only concentrated on differences and potential incompatibilities.

There are obvious many breaking changes coming from `rust-multiaddr`, but that's find and documenting them in depth might not be worth it as current users (if there are any) would need to change a lot. Time might be better spend helping people migrate if they run into issues.

One thing that might be worth mentioning in the changelog/commit message is that you cannot any longer convert from an integer to a `Protocol` and back, though you have constants to get the integer values of protocols.

Other that that I didn't find anything that hasn't a replacement/workaround in the updated version.

mxinden commented on src/protocol.rs:
```diff
@@ -1,499 +1,585 @@
+const UNIX: u32 = 400;
+const UTP: u32 = 302;
+const WS: u32 = 477;
+const WS_WITH_PATH: u32 = 4770;         // Note: not standard
```
From what I know today, yes. Though I still need to dig deeper into the websocket addressing schema to know for sure.

mxinden commented:
> One thing that might be worth mentioning in the changelog/commit message is that you cannot any longer convert from an integer to a Protocol and back, though you have constants to get the integer values of protocols.

Addressed in https://github.com/multiformats/rust-multiaddr/pull/40/commits/11d6fcf5a359c38c101c46665f18b215b844528d.

> It is not Copy any more, it might be worth mentioning in the changelog/commit message.

Addressed in https://github.com/multiformats/rust-multiaddr/pull/40/commits/11d6fcf5a359c38c101c46665f18b215b844528d.


--- 


@vmx thank you for the review. Would you mind taking another look?

vmx reviewed (approved):
Yay! So good to see changes being upstreamed into a single source of truth.

mxinden commented:
> In a few weeks.

@Stebalien I am interpreting this as an offer, not as a request to pause this pull request till then. Let me know if you would prefer  to pause this pull request until you find time to review, after all this effort is in no rush.

---

Unless anything comes up, I will merge this sometime next week and cut the v0.12.0 release. Thanks @vmx  for the help.

Stebalien commented:
> I am interpreting this as an offer, not as a request to pause this pull request till then.

Ah, no. It was an offer to ping me in a few weeks if this got stuck. Thanks @vmx!
